### PR TITLE
feat: [GPR-536][Sale Widget] Get multiple currencies

### PR DESCRIPTION
**Pull request by mimi-imtbl**
# Summary

Add functionality to the sale widget to get multiple currencies. 

## Added 
- Add logic to  `useClientConfig` hook to manage and aggregate currencies 
- Getting lists of currencies from sale BE & checkout config 
- Mark `settlement` vs `swappable` currencies
- Add function to remove duplicated currencies in the aggregated currency list & sort it based on priority: `base` currency -> `settlement` currencies -> `swappable` currencies
- Set `base` currency as `selectedCurrency`
- Add error handling 
- Add tests to utils functions

### CODE REVIEW

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  ClientConfigCurrency,
+  ClientConfigCurrencyConversion,
+  SaleWidgetCurrency,
+  SaleWidgetCurrencyType,
+  SignPaymentTypes,
+} from '../types';
+import { useCheckoutClient } from './useCheckoutClient';
+import { sortCurrencies } from '../utils/sortCurrencies';
+
+type ClientConfigResponse = {
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts:21]: @mimi-imtbl consider applying the renaming consistently across. although i think this hook better keep the prev name as is not specific to currency at alone

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  passport,
+  defaultCurrency,
+}: UseClientConfigParams) => {
+  const [selectedCurrency, setSelectedCurrency] = useState<
+  ClientConfigCurrency | undefined
+  >();
+  const [clientConfig, setClientConfig] = useState<ClientConfig>(defaultClientConfig);
+  const [allCurrencies, setAllCurrencies] = useState<SaleWidgetCurrency[]>([]);
+
+  const checkoutClient = useCheckoutClient({ passport, environment });
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts:84]: We can optimise things here, the checkout instance was already created when the widget instance is created.
Have a look [here](https://github.com/immutable/ts-immutable-sdk/blob/ffd504b90afb04feeb8b08de60ebde8a4dbcbba6/packages/checkout/widgets-lib/src/widgets/BaseWidgetRoot.ts#L22-L22), notice the `BaseWidgetRoot` class provides the checkout instance. so we can pass it down to SaleWidget and then hold it on the Context object.

**Review by jhesgodi — CHANGES_REQUESTED** [review]

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    if (!environment || !environmentId) return;
+
+    const fetchSwappableCurrencies = async () => {
+      if (!checkoutClient || !passport) {
+        return [];
+      }
+      try {
+        const passportProvider = passport.connectEvm();
+        const zkEvmProvider = new providers.Web3Provider(passportProvider);
+        const checkoutNetworkInfo = await checkoutClient.getNetworkInfo({
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts:96]: ideally we should use the already existing provider.

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  currency_conversion: {
+    [key: string]: {
+      amount: number;
+      name: string;
+      type: string;
+    };
+  };
+};
+
+const toClientConfig = (response: ClientConfigResponse): ClientConfig => ({
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts:39]: since this logic became complex, I'll suggest we move into using pattern of creating utility functions within the `/functions` directory. And start adding some test to each utility.

example
https://github.com/immutable/ts-immutable-sdk/blob/853aaee1eb1b00f61d95735576e8b1c8006f949b/packages/checkout/widgets-lib/src/widgets/swap/functions/FetchQuote.ts

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      }
+    };
+
+    (async () => {
+      const [swappableCurrencies, settlementCurrencies] = await Promise.all([
+        fetchSwappableCurrencies(),
+        fetchSettlementCurrencies(),
+      ]);
+
+      const combinedCurrencies: SaleWidgetCurrency[] = [
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts:146]: need to be careful here, we require the final list doesn't have duplicates but also to respect the priority order.
For example, ETH could be returned from the allowlist and `/client-config` in such case it should be taken as swappable

File: packages/checkout/widgets-lib/src/widgets/sale/utils/sortCurrencies.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { SaleWidgetCurrency } from '../types';
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/utils/sortCurrencies.ts:1]: nice one 🔥

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        const config = toClientConfig(data);
+        setClientConfig(config);
+
+        return config.currencies.map((currency) => ({
+          ...currency,
+          currencyType: SaleWidgetCurrencyType.SETTLEMENT,
+        }));
+      } catch (error) {
+        // eslint-disable-next-line no-console
+        console.error('Error fetching settlement currencies:', error);
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts:135]: this may cause the widget to break.
Can you check if throwing here is captured by an error screen?

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts (hunk 2 above)
Comment (mimi-imtbl) [reply packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts]: ah yes! good one!

File: packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts (hunk 7 above)
Comment (mimi-imtbl) [reply packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts]: This is how the `useClientConfig` is currently handling error - a `console.warn` and passing the empty default `clientConfig`. I'll add error handling for it to go straight to the error screen if the error occurs here as transak/coins flow both wouldn't work without `contractId`

File: packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
+import { SaleWidgetCurrency, SaleWidgetCurrencyType } from '../types';
+import { sortAndDeduplicateCurrencies } from './sortAndDeduplicateCurrencies';
+
+describe('sortAndDeduplicateCurrencies', () => {
+  it('sortAndDeduplicateCurrencies', () => {
+    const allCurrencies: SaleWidgetCurrency[] = [
+      {
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts:7]: @mimi-imtbl not a blocker, the input array is already sorted, can you scramble the order to make sure the sort function is actually doing its job?

File: packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+      },
+      {
+        name: 'zkCORE',
+        symbol: 'zkCORE',
+        address: '0x4B96E7b7eA673A996F140d5De411a97b7eab934E',
+        decimals: 18,
+        currencyType: SaleWidgetCurrencyType.SWAPPABLE,
+      },
+      {
+        name: 'USDC',
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts:178]: shouldn't it be deduplicated?

File: packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+      {
+        base: true,
+        decimals: 6,
+        name: 'USDC',
+        address: '0x3b2d8a1931736fc321c24864bceee981b11c3c57',
+        exchangeId: 'usd-coin',
+        currencyType: SaleWidgetCurrencyType.SETTLEMENT,
+      },
+      {
+        base: false,
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts:16]: @mimi-imtbl worth making sure that if 2 bases are returned, it will only work with 1, first one found mostlikely

File: packages/checkout/widgets-lib/src/widgets/sale/functions/transformToClientConfig.test.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+import { ClientConfig, SignPaymentTypes } from '../types';
+import {
+  ClientConfigResponse,
+  transformToClientConfig,
+} from './transformToClientConfig';
+
+describe('transformToClientConfig', () => {
+  it('transformToClientConfig', () => {
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/functions/transformToClientConfig.test.ts:9]: good to describe what being tested 
`it('should return input object with camel case keys')`

File: packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts [hunk 12]
```diff
@@ -0,0 +1,197 @@
+import { SaleWidgetCurrency, SaleWidgetCurrencyType } from '../types';
+import { sortAndDeduplicateCurrencies } from './sortAndDeduplicateCurrencies';
+
+describe('sortAndDeduplicateCurrencies', () => {
+  it('sortAndDeduplicateCurrencies', () => {
```
Comment (jhesgodi) [inline packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts:5]: best to describe what being tested
`it('should remove duplicates and sort currencies by: base -> settlement -> swappable')`

**Review by jhesgodi — APPROVED** [review]
🟢 LGTM, approve with minor comments @mimi-imtbl

File: packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts (hunk 10 above)
Comment (mimi-imtbl) [reply packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts:16]: yep, fixed this edge case

Comment (mimi-imtbl) [reply packages/checkout/widgets-lib/src/widgets/sale/functions/sortAndDeduplicateCurrencies.test.ts:16]: Based on the current implementation, it will pick the later `base: true`. This shouldn't be an issue, I'd assume. I'd prefer to not make this logic more complicated, will leave this for now

**Review by dreamoftrees — APPROVED** [review]
Great work @mimi-imtbl

**Review by jhesgodi — APPROVED** [review]
